### PR TITLE
machines: Allow editing VM's CPU mode and model

### PR DESCRIPTION
**Pull request by KKoukiou**
For the release notes:
### Machines:  Added support for configuring the CPU type

When changing the CPU model, the goals are to maximise performance, while protecting guest OS against various CPU hardware flaws, and optionally enabling live migration between hosts with heterogeneous CPU models.

There are three different options supported.
* Host passthrough

This passes the host CPU model features, model, stepping, exactly to the guest.

* Host model

This uses the QEMU `Named model` feature, automatically picking a CPU model that is similar the host CPU, and then adding extra features to approximate the host model as closely as possible. This does not guarantee the CPU family, stepping, etc will precisely match the host CPU, as they would with `Host passthrough`, but gives much of the benefit of passthrough, while making live migration safe.

* Custom / Named model

QEMU comes with a number of predefined named CPU models, that typically refer to specific generations of hardware released by Intel and AMD. These allow the guest VMs to have a degree of isolation from the host CPU, allowing greater flexibility in live migrating between hosts with differing hardware.

More information about this feature can be found in the [QEMU documentation page](https://www.qemu.org/docs/master/system/qemu-cpu-models.html).

Screen Shot 2020-11-26 at 15 47 11

### CODE REVIEW

**Comment by marusak** [comment]
Needs rebase

File: test/verify/check-machines [hunk 1]
```diff
@@ -2678,6 +2678,36 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
...
+        b.wait_in_text("#vm-subVmTest1-state", "Shut off")
+        # Warning should dissapear and changes should be visible in the UI
+        b.wait_not_present("#cpu-tooltip")
+        b.wait_in_text("#vm-subVmTest1-cpu-model .pf-c-description-list__text", "host")
+
+        # Choose manually a CPU model
+        b.click("#vm-subVmTest1-cpu-model button")
+        b.select_from_dropdown("#cpu-model-select-group select", "coreduo")
+        b.click("#cpu-config-dialog-apply")
+        b.wait_in_text("#vm-subVmTest1-cpu-model .pf-c-description-list__text", "custom (coreduo)")
```
Comment (skobyda) [inline test/verify/check-machines:2709]: Could you try to also start the VM here? This would probably require to add some features to cpu element in domains xml:
```
    <feature policy='disable' name='ss'/>
    <feature policy='disable' name='x2apic'/>
    <feature policy='disable' name='hypervisor'/>
    <feature policy='disable' name='monitor'/>
```

**Review by skobyda — CHANGES_REQUESTED** [review]

**Comment by skobyda** [comment]
Also it seems that everytime I try to change VM do custom cpu, I have to input manually features into an <cpu> in VM XML. To see which features are missing, we can use virsh hypervisor-cpu-compare to see which features are missing, and input them automatically to XML for the use, wdyt?

File: test/verify/check-machines (hunk 1 above)
Comment (marusak) [reply test/verify/check-machines:2709]: It would be great if you could verify that it was set up outside of cockpit - with virt-inspect ot whatever cli tool can show that.

**Comment by KKoukiou** [comment]
> Also it seems that everytime I try to change VM do custom cpu, I have to input manually features into an in VM XML. To see which features are missing, we can use virsh hypervisor-cpu-compare to see which features are missing, and input them automatically to XML for the use, wdyt?

does virt-manager auto-add features?

File: pkg/machines/libvirt-xml-update.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
+import { getElem, getSingleOptionalElem } from './libvirt-common.js';
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-xml-update.js:1]: Makes total sense.

File: test/verify/check-machines (hunk 1 above)
Comment (KKoukiou) [reply test/verify/check-machines:2709]: @skobyda virt-manager has the same bahavior like this PR, at least from my understanding. 
Do you mind if this is true, improvements like this to come up in a followup?

File: pkg/machines/libvirt-xml-update.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-xml-update.js:1]: yes, I usually try to remember to name component files with camelcase, but helpers and other non react files don't have any rule. I don't have a preference here.

[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-xml-update.js:1]: Ah hmm, xmlCreator.js contains functions which contains part of the domain XML. We use them with API functions which take XML and update resources in that way. 
but the libvirt-xml-update function adjust returns always the whole XML of the domain changed.

It's slightly different.

[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-xml-update.js:1]: So since we decided to start using virt-xml I think this file (xmlCreator) will be deprecated soon.
Therefore I will not try to  merge it with something else now.

File: test/verify/check-machines (hunk 1 above)
Comment (KKoukiou) [reply test/verify/check-machines:2710]: virt-xml pupolates there correctly

**Comment by KKoukiou** [comment]
Meh this conficts again now.

File: pkg/machines/components/cpuTypeModal.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                                    setCpuModel(undefined);
+                                } else {
+                                    setCpuModel(value);
+                                    setCpuMode("custom");
+                                }
+                            }}>
+                    <FormSelectOption key="host-model"
+                                      data-value="host-model"
+                                      value="host-model"
+                                      label="host-model" />
```
Comment (marusak) [inline pkg/machines/components/cpuTypeModal.jsx]: shouldn't these labels be localized? I am wondering if they are static values, or they should be translatable (or we should have mapping from them to some human readable forms)?

File: test/verify/check-machines [hunk 4]
```diff
@@ -2710,10 +2710,6 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
         b.click("#cpu-config-dialog-apply")
         b.wait_in_text("#vm-subVmTest1-cpu-model .pf-c-description-list__text", "custom (coreduo)")
 
-        # CPU configuration should be disabled for running VMs
-        b.click("#vm-subVmTest1-run")
-        b.wait_present("#vm-subVmTest1-cpu-model button.pf-m-disabled")
-
```
Comment (marusak) [inline test/verify/check-machines]: Why is this change in commit `machine: Split xml editing function out of libvirt-common `?

File: pkg/machines/components/vm/overview/cpuTypeModal.jsx [hunk 5]
```diff
@@ -19,7 +19,8 @@ export const CPUTypeModal = ({ vm, models, close }) => {
             connectionName: vm.connectionName,
             mode: cpuMode,
             model: cpuModel
-        }).then(close, exc => setError({ dialogError: _("CPU configuration could not be saved"), dialogErrorDetail: exc.message }));
+        })
+                .then(close, exc => setError({ dialogError: _("CPU configuration could not be saved"), dialogErrorDetail: exc.message }));
```
Comment (marusak) [inline pkg/machines/components/vm/overview/cpuTypeModal.jsx:23]: Seems completely unrelated change.

File: pkg/machines/components/vm/overview/vmOverviewCard.jsx [hunk 6]
```diff
@@ -214,13 +220,24 @@ class VmOverviewCard extends React.Component {
...
+            <Button variant="link" className="edit-inline" isInline isAriaDisabled={!vm.persistent || !this.state.virtXMLAvailable} onClick={this.openCpuType}>
+                {_("edit")}
+            </Button>
+        );
+        if (!this.state.virtXMLAvailable) {
+            cpuEditButton = (
+                <Tooltip id='virt-install-missing'
+                         content={_("virt-install package needs to be installed on the system in order to edit this attribute")}>
+                    {cpuEditButton}
+                </Tooltip>
```
Comment (marusak) [inline pkg/machines/components/vm/overview/vmOverviewCard.jsx:171]: I like how much simpler this approach is:)
However, I want to stop for a second and think about this `virt-xml` tool. Is it pulled as dependency by something? Do we expect it to be present in majority of cases? Should we depend on it on our own?

File: pkg/machines/components/vm/overview/bootOrder.jsx [hunk 7]
```diff
@@ -339,4 +341,45 @@ BootOrderModal.propTypes = {
...
+    return bootOrder;
+}
+
+export const BootOrderLink = ({ vm, idPrefix, close, dispatch, nodeDevices }) => {
+    const [bootOrderShow, setBootOrderShow] = useState(false);
+    const bootOrderChanged = () => {
+        const activeDevices = getBootOrderDevices(vm);
+        const inactiveDevices = getBootOrderDevices(vm.inactiveXML);
+
+        // check if number bootable devices has changed
```
Comment (marusak) [inline pkg/machines/components/vm/overview/bootOrder.jsx:367]: "number of bootable" :) (just real nitpick, please feel free to ignore)

Comment (marusak) [reply pkg/machines/components/vm/overview/bootOrder.jsx:367]: just noticed this was just moved from other place, so double ignore :)

File: pkg/machines/app.jsx [hunk 8]
```diff
@@ -22,10 +22,10 @@ import { AlertGroup, Alert } from '@patternfly/react-core';
 import { superuser } from "superuser.js";
```
Comment (marusak) [inline pkg/machines/app.jsx:22]: "machines: bring some order in the direstory structure and file naming " commit message could be fixed.
The title mistypes `directory`, you used `where` instead of `were` twice and spelling of `don't`

**Review by marusak — COMMENTED** [review]
Nice!
I really don't have strong options about naming and directory structure as I am not familiar enough with the code. Reorg changes make sense and I trust your decision :)

I commented some mostly nitpicks like typos in comments, commits or unrelated changes in files or in commits. Nothing of this is real issue. I haven't found anything functionally incorrect.

File: pkg/machines/components/cpuTypeModal.jsx (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/components/cpuTypeModal.jsx]: I am not sure if we want to localize these, these are attributes of the XML configuration and I think libvirt XML attributes should not be localized so that users can understand the mapping between our dialogs and libvirt XML files.
https://libvirt.org/formatdomain.html#cpu-model-and-topology
in the case the mapping it to:
``` xml
<cpu mode='host-model'>
```

``` xml
<cpu mode='host-passthrough'> 
```

File: test/verify/check-machines (hunk 4 above)
Comment (KKoukiou) [reply test/verify/check-machines]: Right sorry:
moved that to: commit a5fc888c90c6bcfa0f60f59d59c904c18885d612
Author: Katerina Koukiou <kkoukiou@redhat.com>
Date:   Mon Nov 30 11:24:31 2020 +0100

    machines: Allow editing cpu mode/model even when the VM is running

File: pkg/machines/components/vm/overview/cpuTypeModal.jsx (hunk 5 above)
Comment (KKoukiou) [reply pkg/machines/components/vm/overview/cpuTypeModal.jsx:23]: Indeed, removed it.

File: pkg/machines/components/vm/overview/vmOverviewCard.jsx (hunk 6 above)
Comment (KKoukiou) [reply pkg/machines/components/vm/overview/vmOverviewCard.jsx:171]: virt-xml is part of virt-install packagel. no extra dependency is needed. virt-install is already a recommeneded dependenncy https://github.com/cockpit-project/cockpit/blob/master/tools/cockpit.spec#L655

Users can already use cockpit-machines without having virt-install package installed, but these will cause them to not be able to create new VMs. Now they won't be able to edi CPU model either, they would see the tooltip prompting them to install virt-install. 

I think it's a fair trade off.
Pro:
----- 
One one side virt-xml will simplyfy rastically our code removing all XML editing functions
 (especially once I port all XML editing to this tool)

I expect it to be more error phrone, it's a tool shipped by the virt-team which will replace my/Simon's code which can always contain bugs. 

virt-xml is much more sophisticated than our dummy XML creation, there is a lot of tweaks that can be done for increasing VM performance which I never implement because I am not familiar

Cotra:
-------- 
 Some users which decide to not install 'virt-install', they will just be able to perform lifecycle operations on existing XMs but not really create new ones or edit the configuration.

File: pkg/machines/app.jsx (hunk 8 above)
Comment (KKoukiou) [reply pkg/machines/app.jsx:22]: that was horrible

File: pkg/machines/components/cpuTypeModal.jsx (hunk 3 above)
Comment (marusak) [reply pkg/machines/components/cpuTypeModal.jsx]: Fair enough.

File: pkg/machines/components/vm/overview/vmOverviewCard.jsx (hunk 6 above)
Comment (marusak) [reply pkg/machines/components/vm/overview/vmOverviewCard.jsx:171]: Cool, if it is part of virt-install and also it is recommended by us, then this is surely fine :) The amount of installations where users would opt out is going to be minimal and since you show nice warning about this, I think this is completely fine and good approach!

**Review by marusak — APPROVED** [review]
Thanks!
